src/ripple/app/misc/NegativeUNLVote.cpp

```diff
@@ -0,0 +1,354 @@
+        obj.setFieldU32(sfLedgerSequence, seq);
+        obj.setFieldVL(sfUNLModifyValidator, vp.slice());
+    });
+
+    uint256 txID = nUnlTx.getTransactionID();
+    Serializer s;
+    nUnlTx.add(s);
+    auto tItem = std::make_shared<SHAMapItem>(txID, s.peekData());
+    if (!initialSet->addGiveItem(tItem, true, false))
+    {
+        JLOG(j_.warn()) << "N-UNL: ledger seq=" << seq
+                        << ", add ttUNL_MODIFY tx failed";
+    }
+    else
+    {
+        JLOG(j_.debug()) << "N-UNL: ledger seq=" << seq
+                         << ", add a ttUNL_MODIFY Tx with txID: " << txID
+                         << ", the validator to "
+                         << (disabling ? "disable: " : "re-enable: ") << vp;
+    }
+}
+
+NodeID
+NegativeUNLVote::pickOneCandidate(
+    uint256 randomPadData,
+    std::vector<NodeID>& candidates)
+{
+    assert(!candidates.empty());
+    static_assert(NodeID::bytes <= uint256::bytes);
+    NodeID randomPad = NodeID::fromVoid(randomPadData.data());
+    NodeID txNodeID = candidates[0];
+    for (int j = 1; j < candidates.size(); ++j)
+    {
+        if ((candidates[j] ^ randomPad) < (txNodeID ^ randomPad))
+        {
+            txNodeID = candidates[j];
+        }
+    }
+    return txNodeID;
+}
+
+bool
+NegativeUNLVote::buildScoreTable(
+    LedgerConstPtr& prevLedger,
+    hash_set<NodeID> const& unl,
+    RCLValidations& validations,
+    hash_map<NodeID, unsigned int>& scoreTable)
+{
+    assert(scoreTable.empty());
+
+    /*
+     * Find agreed validation messages received for the last 256 ledgers,
+     * for every validator, and fill the score table.
+     *
+     * -- ask the validation container to keep enough validation message
+     *    history for next time.
+     *
+     * -- find 256 previous ledger hashes, the same length as a flag ledger
+     *    period.
+     *
+     * -- query the validation container for every ledger hash and fill
+     *    the score table.
+     *
+     * -- return false if the validation message history or local node's
+     *    participation in the history is not good.
+     */
+
+    auto const seq = prevLedger->info().seq + 1;
+    validations.setSeqToKeep(seq - 1);
+
+    auto const hashIndex = prevLedger->read(keylet::skip());
+    if (!hashIndex || !hashIndex->isFieldPresent(sfHashes))
+    {
+        JLOG(j_.debug()) << "N-UNL: ledger " << seq << " no history.";
+        return false;
+    }
+
+    auto ledgerAncestors = hashIndex->getFieldV256(sfHashes).value();
+    auto numAncestors = ledgerAncestors.size();
+    if (numAncestors < 256)
+    {
+        JLOG(j_.debug()) << "N-UNL: ledger " << seq
+                         << " not enough history. Can trace back only "
+                         << numAncestors << " ledgers.";
+        return false;
+    }
+
+    // have enough ledger ancestors, build the score table
+    for (auto const& k : unl)
+    {
+        scoreTable[k] = 0;
+    }
+    auto idx = numAncestors - 1;
```

Comment:
I don't think we need the `idx` variable. Just do `numAncestors -i -1` as the index.

Reply:
Thanks, changed.